classes/village/VillageUpgradeManager.php

```diff
@@ -134,55 +147,82 @@ public static function initializeEffectsForVillage(System $system, array $upgrad
-                        description: VillageUpgradeConfig::UPGRADE_DESCRIPTIONS[$upgrade_key],
-                        materials_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_MATERIALS],
-                        food_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_FOOD],
-                        wealth_research_cost: VillageUpgradeConfig::UPGRADE_RESEARCH_COST[$upgrade_key][WarManager::RESOURCE_WEALTH],
-                        research_time: VillageUpgradeConfig::UPGRADE_RESEARCH_TIME[$upgrade_key],
-                        research_time_remaining: isset($village->upgrades[$upgrade_key]->research_progress_last_updated) ? System::TimeRemaining($village->upgrades[$upgrade_key]->research_progress_required - $village->upgrades[$upgrade_key]->research_progress, format: "long", include_seconds: false, include_minutes: false) : null,
-                        food_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_FOOD],
-                        materials_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_MATERIALS],
-                        wealth_upkeep: VillageUpgradeConfig::UPGRADE_UPKEEP[$upgrade_key][WarManager::RESOURCE_WEALTH],
+                        status: $upgrade ? $upgrade->status : VillageUpgradeConfig::UPGRADE_STATUS_LOCKED,
+                        research_progress: $upgrade ? $upgrade->research_progress : null,
+                        research_progress_required: $upgrade ? $upgrade->research_progress_required : null,
+                        research_progress_last_updated: $upgrade ? $upgrade->research_progress_last_updated : null,
+                        research_boosted: $upgrade ? $upgrade->research_boosted : false,
+                        research_time_remaining: $upgrade ? System::TimeRemaining($upgrade->research_progress_required - $upgrade->research_progress, format: "long", include_seconds: false, include_minutes: false) : '',
+                        name: $upgrade_config_data->getName(),
+                        tier: $upgrade_config_data->getTier(),
+                        description: $upgrade_config_data->getDescription(),
+                        materials_research_cost: ($upgrade && $upgrade->research_progress !== null) ? 0 : $upgrade_config_data->getResearchCostMaterials(),
+                        food_research_cost: ($upgrade && $upgrade->research_progress !== null) ? 0 : $upgrade_config_data->getResearchCostFood(),
+                        wealth_research_cost: ($upgrade && $upgrade->research_progress !== null) ? 0 : $upgrade_config_data->getResearchCostWealth(),
+                        research_time: ($upgrade && $upgrade->research_progress !== null) ? round($upgrade_config_data->getResearchTime() * (($upgrade->research_progress_required - $upgrade->research_progress) / $upgrade->research_progress_required)) : $upgrade_config_data->getResearchTime(),
```

Comment:
this formatting makes it easier to read long ternaries
```suggestion
                        research_time: ($upgrade && $upgrade->research_progress !== null) 
                            ? round($upgrade_config_data->getResearchTime() * (($upgrade->research_progress_required - $upgrade->research_progress) / $upgrade->research_progress_required)) 
                            : $upgrade_config_data->getResearchTime(),
```

Reply:
I think I'll even make this a separate calculation before the class constructor.